pkg/apis/crd/v1alpha2/types.go

```diff
@@ -543,3 +543,94 @@ type TrafficControlList struct {
 
 	Items []TrafficControl `json:"items"`
 }
+
+type BGPPolicyStatus string
+
+const (
+	BGPPolicyActive  = "Active"
+	BGPPolicyStandby = "Standby"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +genclient:noStatus
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// BGPPolicy defines BGP configuration applied to Nodes.
+type BGPPolicy struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard metadata of the object.
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec BGPPolicySpec `json:"spec"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+type BGPPolicyList struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty"`
+
+	Items []BGPPolicy `json:"items"`
+}
+
+// BGPPolicySpec defines the specification for a BGP policy.
+type BGPPolicySpec struct {
+	// NodeSelector is utilized to select Nodes to which the policy is applied.
+	NodeSelector *metav1.LabelSelector `json:"nodeSelector"`
+
+	// LocalASN is the AS number used by BGP protocol.
+	LocalASN int16 `json:"localASN"`
+
+	// BindAddresses is a list of IP addresses binding to BGP protocol.
+	BindAddresses []string `json:"bindAddresses,omitempty"`
+
+	// ListenPort is the port on which BGP protocol listen.
+	ListenPort *int16 `json:"listenPort,omitempty"`
+
+	// RouterID is the routerID used by BGP protocol in the format of an IPv4 address. This field is required on an IPv6
+	// single stack cluster.
+	RouterID *string `json:"routerID,omitempty"`
+
+	// Advertisements selects routes to be advertised to external BGP peers.
+	Advertisements Advertisements `json:"advertisements"`
+
+	// BGPPeers are the list of external BGP peers.
+	BGPPeers []BGPPeer `json:"bgpPeers"`
+}
+
+type Advertisements struct {
+	// Determine whether to advertise ClusterIPs.
+	ClusterIPs bool `json:"clusterIPs,omitempty"`
```

Comment:
I dropped earlier and did not catch the discussion in the community meeting. When we support Service/Pod selectors later, how are we going to extend the API?

Reply:
The main issue I can see is if we want to advertise different types of IPs for different Services. Do we think this is a valid use case?

We also end up with a few different ways of expressing the same thing, which is not great. For example, looking at the proposed `PodIPs *PodAdvertisement` field, the following are equivalent:
```
PodIPs: nil 
```
```
PodIPs: &PodAdvertisement{}
```

It feels like if the long-term plan is to add selectors, a more natural API would be:
```go
type Advertisements struct {
    Pods *PodAdvertisement
    // OR
    Pods []PodAdvertisement
}
type PodAdvertisement {
    PodSelector *metav1.LabelSelector
    NamespaceSelector  *metav1.LabelSelector
}
```
In that case a `nil` `Pods` field would mean "don't advertise Pod IPs", an empty `&PodAdvertisement{}` would mean "advertise all Pod IPs", and providing selectors would filter which Pod IPs need to be advertised.

But maybe there is some readability value in having an explicit boolean.

Reply:
@jianjuns do we see any long term use case for advertising secondary Pod IPs?

Reply:
> The main issue I can see is if we want to advertise different types of IPs for different Services. Do we think this is a valid use case?

Maybe we could advertise `ExternalIPs` and `LoadBalancerIPs` in the first version, since it seems more meaningful to advertise them, and it is easy to add `ClusterIPs` if it is really needed in the future.

Reply:
> We also end up with a few different ways of expressing the same thing, which is not great. For example, looking at the proposed PodIPs *PodAdvertisement

You are right. This reminded me. Different ways of expressing the same thing is an issue.  Since we don't add selectors currently, the struct will be:

```go
type Advertisements struct {
    Pods *PodAdvertisement
    // OR
    Pods []PodAdvertisement
}
type PodAdvertisement {}
```
Personally, it might confuse the users that want to advertise all Pod IPs like this `PodIPs: &PodAdvertisement{}` instead of a boolean type. However, to leave the space to extend in the future, maybe we have to do so.

Reply:
If we already know that we want to add selector later on, it does feel like we should accommodate for that, even if the API ends up looking "strange" in the first version.
The simplest solution would be:
```go
type Advertisements struct {
    Pods []PodAdvertisement
    Services []ServiceAdvertisement
}
type PodAdvertisement struct {
    // empty now, selectors to be added later
    // PodSelector *metav1.LabelSelector
    // NamespaceSelector  *metav1.LabelSelector
}
type ServiceAdvertisement struct {
   ClusterIPs bool
   LoadBalancerIPs bool
   ExternalIPs bool
   // selectors to be added later
   // PodSelector *metav1.LabelSelector
   // NamespaceSelector  *metav1.LabelSelector
}
```
I like having a list (`Pods []PodAdvertisement`) instead of a pointer (`Pods *PodAdvertisement`) as I feel like it makes it easier to select multiple independent workloads.

What's a bit unusual:
- the absence of selector will mean "select all"
- until we introduce selectors, the API looks unintuitive

To select all Services:
```yaml
advertisements:
    services:
        - clusterIPs: false
        - loadBalancerIPs: true
        - externalIPs: true
```
To select all Pods:
```yaml
advertisements:
    pods:
        - {}
```

Another option:
```go
type BGPPolicy struct {
  // ...
  Advertisements []Advertisement
}

type Advertisement struct {
  Pods *PodAdvertisement
  Services *ServiceAdvertisement
}

// same definition for PodAdvertisement & ServiceAdvertisement
```
IMO, this option is less natural.

We should try to close on the API design soon.

Reply:
+1 for @antoninbas 's first option. I do not feel it is ideal either, but do not have a better idea.

Reply:
Thanks for @antoninbas 's detailed explanation. I prefer the first option as well. One more thing about the selector: do we all agree on using a boolean type to select Egress IPs? If so, the struct should be:

```go
type Advertisements struct {
    Pods []PodAdvertisement
    Services []ServiceAdvertisement
    Egresses bool
}
```

Reply:
I don't remember exactly all the discussions we had, but I don't think we need a selector for Egresses. We also didn't see a request for this. An Egress is a global (cluster-scope) resource. If one only wanted to advertise a subset of Egress IPs, they could split the corresponding ExternalIPPools across 2 sets of Nodes, and configure BGPPolicies for these Nodes accordingly.  cc @jianjuns @tnqn 

I would prefer to name the field `EgressIPs` though.

```go
type Advertisements struct {
    Pods []PodAdvertisement
    Services []ServiceAdvertisement
    EgressIPs bool
}
```

If we need to add more complex selection for Egresses later one, we will be able to add `Egresses []EgressAdvertisement` and deprecate `EgressIPs bool`. But even without this consideration, I have a slight preference for `EgressIPs`, which indicates that it is defined differently from the other 2 fields.

Reply:
For consistency, still good to define a struct for Egress IPs? Maybe it is possible to advertise to different routers for different Egresses on different Node pools?

Reply:
> For consistency, still good to define a struct for Egress IPs?

But a list `[]EgressAdvertisement` would not make sense here IMO

>  Maybe it is possible to advertise to different routers for different Egresses on different Node pools?

Isn't this achievable with a simple boolean and a different BGPPolicy (with different peers) for each Node pool? This seems to be similar to what I was trying to describe above as well:

> If one only wanted to advertise a subset of Egress IPs, they could split the corresponding ExternalIPPools across 2 sets of Nodes, and configure BGPPolicies for these Nodes accordingly

Reply:
> Isn't this achievable with a simple boolean and a different BGPPolicy (with different peers) for each Node pool?
This is true. I am fine with a boolean then. Maybe @tnqn has an opinion too.

Reply:
Regarding consistency and extensibility, how about defining the first level fields in the same way and customize each kind of resources in their own structs (even it's empty for now):
```go
type Advertisements struct {
	Pod     *PodAdvertisement
	Service *ServiceAdvertisement
	Egress  *EgressAdvertisement
}

type PodAdvertisement struct {
	// empty now, selectors to be added later
	// Selectors  []Selector
}

type ServiceAdvertisement struct {
	ClusterIPs      bool
	LoadBalancerIPs bool
	ExternalIPs     bool
	// empty now, selectors to be added later
	// Selectors  []Selector
}

type EgressAdvertisement struct {
	// empty now, selectors to be added later
	// Selectors  []Selector
}
```
Some example configurations:
* Advertise Pod and Egress IPs:
```
advertisement:
  pod: {}
  egress: {}
```
* Advertise Service LoadBalancerIPs:
```
advertisement:
  service:
    loadBalancerIPs: true
```
* Advertise specific Services' ClusterIPs and specific Pod IPs (when/if selector is supported):
```
advertisement:
  service:
    clusterIPs: true
    selectors: 
    - namespaceSelector: 
        foo: bar
      podSelector:
        foo: bar
    - namespaceSelector:
        foo: baz
  pod:
    selectors: 
    - namespaceSelector: 
        foo: bar
      podSelector:
        foo: bar
```

It shouldn't be hard for K8s users to understand that `nil` represents this part will be disabled as it's widely used in K8s APIs.

Reply:
Works for me, even though we are unlikely to need actual selectors for Egress resources.
We will probably need a more feature-specific name for the `Selector` struct.

We cannot choose to advertise different Service address types for different Services with the latest proposal, but AFAIK we don't have an actual requirement for this?

Reply:
With this proposal, if I need multiple Selectors to select Services, I need to create multiple BGPPolicies? I think @antoninbas raised this in his earlier comment.

Reply:
Antonin reminded me the idea is to use a Selector slice, but does that mean it is a shared Selector struct for Pods and Services. Is that good? @tnqn

Reply:
> We will probably need a more feature-specific name for the Selector struct.

Yes, something like the following?
```go
type BGPSelector struct {
	ObjectSelector    *metav1.LabelSelector
	NamespaceSelector *metav1.LabelSelector
}
```

> We cannot choose to advertise different Service address types for different Services with the latest proposal, but AFAIK we don't have an actual requirement for this?

Yes, I even think the three types of IPs won't be advertised together:
* The only use case of advertising clusterIP I'm aware of is when users don't have externally-reachable Service IPs.
* I think ExternalIP and LBIP are kind of mutually exclusive from usage's perspective?

> Antonin reminded me the idea is to use a Selector slice, but does that mean it is a shared Selector struct for Pods and Services. Is that good? @tnqn

Regardless of shared Selector struct or one Selector struct per resource, it will be agnostic to users and will not affect compatibility when merging multiple to one or the opposite, so I'm fine with either but would prefer a shared one when they are exactly same to save some code.

Reply:
Is it an existing pattern to use a generic `ObjectSelector `?

Reply:
I see both patterns in different APIs, but I feel generic structs are more common, for example, see the usage of `ObjectReference` and `LocalObjectReference` in https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/types.go

Reply:
How about Selectors? I do not see ObjectSelector before.

Reply:
You mean the name of the Selector struct? I thought @antoninbas suggested using a name specific to the feature, so I was thinking something like `BGPSelector`. `ObjectSelector` was only used as the field name that selects objects in Namespaces. If we want to make the Selector struct shared among Services and Pods, the field can't be ServiceSelector or PodSelector. I see `ObjectSelector` is used for this purpose in some K8s APIs: https://github.com/kubernetes/kubernetes/blob/f4e246bc93ffb68b33ed67c7896c379efa4207e7/pkg/apis/admissionregistration/types.go#L832.
I suppose you mean the following?
```
type Selectors struct {
	ObjectSelector    *metav1.LabelSelector
	NamespaceSelector *metav1.LabelSelector
}
```

Reply:
Yes, I meant a generic ObjectSelector that can select different resources based on where it is used. Is it a common pattern in K8s?

Reply:
I think `MatchResources` in `ValidatingAdmissionPolicy` in the above link is an example, it selects any kind of resource via `ObjectSelector` and `NamespaceSelector`. There aren't many other examples because they typically select one kind of resource, and the selector normally has a prefix specific to their targets like podSelector, nodeSelector. But we can also find the generic Reference struct `ObjectReference` and `LocalObjectReference` are widely used to reference any kind of object.

Reply:
I feel our usage is even vague in that the type of resources are selected depends on where the Selector is used. Not sure if that is a good API definition.

Reply:
I think it's not very different from the usage of `LocalObjectReference` like the folllowing, in which it represents a reference to ConfigMap in the context of ConfigMapVolumeSource while it represents a refernce to Secret in the context of SecretKeySelector?
```
type ConfigMapVolumeSource struct {
	LocalObjectReference
	Items []KeyToPath
	DefaultMode *int32
	Optional *bool
}

type SecretKeySelector struct {
	LocalObjectReference
	Key string
	Optional *bool
}
```

Reply:
I have updated the code as @tnqn 's latest review comments. Could we have some discussions about somewhere else in this PR?

Reply:
I didn't make it clear. I was just wondering if you could review other fields of the API in this PR.